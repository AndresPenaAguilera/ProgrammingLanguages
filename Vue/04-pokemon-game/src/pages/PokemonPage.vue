<template>
    <div>
    <h1>¿Quién es este pokémon?</h1>
        
    </div>

    <div v-if="!pokemon">Espere por favor...</div>
    <div v-else>
        <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
        <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer"/>
        
        <div v-if="showAnswer">
        <h2 class="fade-in">{{ message }}</h2>
        <button @click="newGame">
            Nuevo juego
        </button>
        </div>
       
    </div>




</template>

<script>

import PokemonOptions from '@/components/PokemonOptions'
import PokemonPicture from '@/components/PokemonPicture'

import getPokemonOptions from '@/helpers/getPokemonOptions'



export default {
    components:{ PokemonOptions, PokemonPicture, getPokemonOptions },
    data(){
        return {
            pokemonArr: [],
            pokemon :null,
            showPokemon : false,
            message: null,
            showAnswer : false
        }
    },
    methods:{
        async mixPokemonArray(){
            this.pokemonArr = await getPokemonOptions()
            
            const randonInt = Math.floor( Math.random() * 4)

            this.pokemon  = this.pokemonArr[ randonInt ]

        },
        checkAnswer( pokemonId ){
            
            this.showPokemon = true
            this.showAnswer = true

            const { id, name } = this.pokemon

            this.message = (id==pokemonId)?`Correcto, ${name}`: `Oops era ${name}`
           

        },
        newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr =[]
            this.pokemon = null
            this.mixPokemonArray() 
        }
    },
    mounted(){
        this.mixPokemonArray()
    }
}
</script>

