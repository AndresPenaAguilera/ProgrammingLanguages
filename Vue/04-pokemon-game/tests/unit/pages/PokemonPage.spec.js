
import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage', ()=>{

    let wrapper
    beforeEach(()=>{
        wrapper = shallowMount( PokemonPage )

    })

    test('debe de llamar el mixPokemonArray al montar', ()=>{

        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray')
        wrapper = shallowMount( PokemonPage )
        expect( mixPokemonArraySpy ).toHaveBeenCalled()


    })


    test('debe de hacer match con el snapshot cuando cargan los componentes', ()=>{

        const wrapper = shallowMount( PokemonPage,{
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon :pokemons[0],
                    showPokemon : false,
                    message: '',
                    showAnswer : false
                }
            }
        })


        expect( wrapper.html() ).toMatchSnapshot()


    })

    test('debe mostrar los componentes de PokemonPicture y PokemonOptions', ()=>{

        const wrapper = shallowMount( PokemonPage,{
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon :pokemons[0],
                    showPokemon : false,
                    message: '',
                    showAnswer : false
                }
            }
        })

        const pokemonPicture = wrapper.find('pokemon-picture-stub')

        expect( pokemonPicture.exists() ).toBeTruthy()
        expect( pokemonPicture.attributes('pokemonid') ).toBe("1")

        const pokemonOptions = wrapper.find('pokemon-options-stub')

        expect( pokemonOptions.exists() ).toBeTruthy()
        expect( pokemonOptions.attributes('pokemons') ).toBeTruthy()
        
    })



    test('debe checkAnswer validar estado the seleccion', async()=>{

        const wrapper = shallowMount( PokemonPage,{
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon :pokemons[0],
                    showPokemon : false,
                    message: '',
                    showAnswer : false
                }
            }
        })


       await wrapper.vm.checkAnswer(1)

       expect( wrapper.find('h2').exists() ).toBeTruthy()
       expect( wrapper.find('h2').text() ).toBe(`Correcto, ${ pokemons[0].name }`)

       await wrapper.vm.checkAnswer(10)

       expect( wrapper.vm.message ).toBe(`Oops era ${ pokemons[0].name }`)

    })


})