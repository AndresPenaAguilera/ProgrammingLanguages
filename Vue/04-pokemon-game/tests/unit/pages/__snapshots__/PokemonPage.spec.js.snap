// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage debe de hacer match con el snapshot cuando cargan los componentes 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
</div>
<div>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
